### PR TITLE
Changed ServerCompositor.cs to unsub from RenderLoop

### LINKED ISSUE #7608: Rendering engine issues High GPU and CPU when app started minimized or idling
Something is not working well with the rendering engine, i will explain.
There are two separate issues that both affect CPU and GPU usage:

1. When the window is open the GPU goes up to 6%-7% and the CPU up to 2% without any activity or UI updates.
2.  When starting an application minimized the GPU and CPU usage are high without justification.

About (1) there is nothing to elaborate, those stats are too high.
regarding (2):
When starting the application normally and than minimizing GPU and CPU are at 0%, as expected,

Now take a look at the following screen shots.

Starting the application minimized (window is not visible):
image
CPU can go up to 2% and GPU to 1.5%

Window is visible:
image

Window minimized again:
image

### PR DESCRIPTION
**Pull request by adirh3**
## What does the pull request do?
This aims to make Avalonia apps have zero idle CPU usage when all windows are not visible or closed.

## What is the current behavior?
Currently Avalonia does not stop the internal render loop when all windows are not visible or even closed.
This caused (low) idle CPU usage, that can be noticeable by users, especially on laptops.
image

## What is the updated/expected behavior with this PR?
Zero idle CPU usage when all windows are minimized/closed on Windows.
image
To check this, simply run the control catalog and minimize the window.
You should see zero CPU usage only when it's minimized (depending on CPU you might not see 0.00X on Task Manager, Process Hacker shows it).

## How was the solution implemented (if it's not obvious)?

There is already support for stopping the `IRenderTimer` through the `Tick` event subscription, but the two main subscribers - `RenderLoopClock` and `ServerCompositor` are not unsubscribing from it. 
I changed both to do that, but I am not quite sure about the `RenderLoopClock` @kekekeks would be nice if you can look at it.

The second issue is that `WinUICompositorConnection` which is the Windows 10/11 `IRenderTimer` never stop working, even if there are zero subscriptions to its `Tick` event.
I changed it to wait using `Semaphore` until there is at least one subscription before it starts it's rendering loop.

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Fixed issues

Fixes #7608

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0026602-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by hez2010** [comment]
I don't think this is feasible for every user. Some programs rely on the render timer to get the correct behavior (especially games), should we provide an option to opt-out this behavior?

File: src/Windows/Avalonia.Win32/WinRT/Composition/WinUICompositorConnection.cs [hunk 1]
```diff
@@ -124,15 +128,25 @@ private void RunLoop()
...
             AppDomain.CurrentDomain.ProcessExit += (sender, args) =>
                 cts.Cancel();
-                
+            var asyncActionCompletedHandler = new RunLoopHandler(this);
             using (var act = _compositor5.RequestCommitAsync())
-                act.SetCompleted(new RunLoopHandler(this));
+            {
+                act.SetCompleted(asyncActionCompletedHandler);
+            }
+            _semaphore.Wait();
```
Comment (kekekeks) [inline src/Windows/Avalonia.Win32/WinRT/Composition/WinUICompositorConnection.cs:136]: I believe that code with `ManualResetEvent` would be less complicated

**Review by kekekeks — CHANGES_REQUESTED** [review]

File: src/Avalonia.Base/Rendering/Composition/CompositingRenderer.cs [hunk 2]
```diff
@@ -295,10 +295,17 @@ public void Paint(Rect rect)
             CompositionTarget.ImmediateUIThreadRender();
     }
 
-    public void Start() => CompositionTarget.IsEnabled = true;
+    public void Start()
+    {
+        if (!CompositionTarget.IsEnabled)
+            Compositor.Server.Start();
```
[the remark this answers is not in the document]
Comment (kekekeks) [reply src/Avalonia.Base/Rendering/Composition/CompositingRenderer.cs:301]: See `AddCompositionTarget`/`RemoveCompositionTarget`

**Comment by kekekeks** [comment]
Also note that `ServerCompositor` _has_ to process any pending composition batches even if there are no active windows. Not doing so can trigger potential memory leak issues.

**Comment by kekekeks** [comment]
Also note that both `Compositor` and `CompositionTarget` are public APIs that can be used by user code outside of `CompositingRenderer`.

File: src/Windows/Avalonia.Win32/WinRT/Composition/WinUICompositorConnection.cs (hunk 1 above)
Comment (adirh3) [reply src/Windows/Avalonia.Win32/WinRT/Composition/WinUICompositorConnection.cs]: Thanks

File: src/Avalonia.Base/Rendering/Composition/CompositingRenderer.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (adirh3) [reply src/Avalonia.Base/Rendering/Composition/CompositingRenderer.cs]: When I put it there I saw the `CompositionTarget` methods are called from the render thread, which in WinUI case is different one.
I moved it there now with `Dispatcher.Post`, which I am not sure if can cause threading issues as well. LMK what you think

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0026936-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-nightly/nuget/v3/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0026937-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-nightly/nuget/v3/index.json) [PRBUILDID]

**Comment by adirh3** [comment]
> I don't think this is feasible for every user. Some programs rely on the render timer to get the correct behavior (especially games), should we provide an option to opt-out this behavior?

Even when the app (or game) minimized? I'm not sure i understand this scenario, but if the app subscribes to the render timer it will not stop. 
This will kick in only once all the subscribers of `IRenderTimer` will unsubscribe from the `Tick` event.

**Comment by adirh3** [comment]
@kekekeks marked this as ready for review but haven't fixed merge conflicts yet. Does #9639 somehow address this issue?

**Comment by kekekeks** [comment]
#9639 only affects graphics contexts and fixes some "Collection was modified" errors in the render loop

File: src/Avalonia.Base/Rendering/Composition/CompositingRenderer.cs [hunk 3]
```diff
@@ -295,7 +295,11 @@ public void Paint(Rect rect)
             CompositionTarget.ImmediateUIThreadRender();
     }
 
-    public void Start() => CompositionTarget.IsEnabled = true;
+    public void Start()
+    {
+        CompositionTarget.IsEnabled = true;
+        CompositionTarget.ImmediateUIThreadRender();
```
Comment (kekekeks) [inline src/Avalonia.Base/Rendering/Composition/CompositingRenderer.cs:301]: This is not guaranteed to be supported on all platforms.

Comment (adirh3) [reply src/Avalonia.Base/Rendering/Composition/CompositingRenderer.cs:300]: @kekekeks any suggestions what to do then?

Comment (adirh3) [reply src/Avalonia.Base/Rendering/Composition/CompositingRenderer.cs]: @kekekeks I believe I fixed it, moved the check to every place before the `Commit().Wait()` by checking the `IsEnabled` property of the `ServerCompositionTarget`.
This should make sure there are no deadlocks, but I am still not familiar with other platforms, so I will need your feedback on that.

**Comment by adirh3** [comment]
@kekekeks I am closing this PR, I don't have enough knowledge of other OS thread rendering behaviors and it's already as risky in Windows with WinUI enabled (many potentials deadlocks).
If you can pick up this PR it would be appreciated, I believe this capability is expected from a lightweight framework like Avalonia.

**Comment by kekekeks** [comment]
@adirh3 
I'll take a shot at subscription management sometime closer to the 11.0 release date.
We now also have some plans to rework the way animations are triggered, so they don't depend on the render timer at all.

Thank you for the investigation and the prototype, those would come useful.

**Comment by adirh3** [comment]
@kekekeks Just wondering - what's your take on this after 11.0 release?
Do you think implementing subscription management for invisible windows is more manageable now?

**Pull request closed without merging**